subsys/net/buf.c

```diff
@@ -91,23 +79,161 @@ void net_buf_reset(struct net_buf *buf)
 	NET_BUF_ASSERT(buf->flags == 0);
 	NET_BUF_ASSERT(buf->frags == NULL);
 
-	buf->len   = 0;
-	buf->data  = buf->__buf;
+	net_buf_simple_reset(&buf->b);
+}
+
+static u8_t *generic_data_ref(struct net_buf *buf, u8_t *data)
+{
+	u8_t *ref_count;
+
+	ref_count = data - 1;
+	(*ref_count)++;
+
+	return data;
+}
+
+static u8_t *mem_pool_data_alloc(struct net_buf *buf, size_t size,
+				 s32_t timeout)
+{
+	struct net_buf_pool *buf_pool = net_buf_pool_get(buf->pool_id);
+	struct k_mem_pool *pool = buf_pool->alloc->alloc_data;
+	struct k_mem_block block;
+	u8_t *ref_count;
+
+	size += sizeof(struct k_mem_block_id) + 1;
+	if (k_mem_pool_alloc(pool, &block, size, timeout)) {
+		return NULL;
+	}
+
+	/* save the block descriptor info at the start of the actual block */
+	memcpy(block.data, &block.id, sizeof(block.id));
+
+	ref_count = (u8_t *)block.data + sizeof(block.id);
+	*ref_count = 1;
+
+	/* Return pointer to the byte following the ref count */
+	return ref_count + 1;
```

Comment:
The net_buf data pointer has never carried with it a promise that it follows a certain alignment. But you're right that some users of net_buf may benefit from it. The simplest solution (not requiring any change to the net_buf implementation) is for the user to use net_buf_reserve() to increment the data pointer until it falls on the desired alignment boundary. Another option is to put the ref count at the end of the data instead of the beginning.